iroh-resolver/src/balanced_tree.rs

```diff
@@ -0,0 +1,380 @@
+            if tree[0].len() == max_degrees {
+                // if so, iterater through nodes
+                for i in 0..tree_len {
+                    // if we encounter any nodes that are not full, break
+                    if tree[i].len() < max_degrees {
+                        break;
+                    }
+
+                    // in this case we have a full set of links & we are
+                    // at the top of the tree. Time to make a new layer.
+                    if i == tree_len - 1 {
+                        tree.push_back(vec![]);
+                    }
+
+                    // create node, keeping the cid
+                    let links = std::mem::replace(&mut tree[i], Vec::new());
+                    let (cid, bytes) = TreeNode::Stem(links).encode()?;
+                    let len = bytes.len();
+                    yield (cid, bytes);
+
+                    // add cid to parent node
+                    tree[i+1].push((cid, len));
+                }
+                // at this point the tree will be able to recieve new links
+                // without "overflowing", aka the leaf node and stem nodes
+                // have fewer than max_degrees number of links
+            }
+
+            // now that we know the tree is in a "healthy" state to
+            // recieve more links, add the link to the tree
+            let (cid, bytes) = TreeNode::Leaf(chunk).encode()?;
+            let len = bytes.len();
+            tree[0].push((cid, len));
+            yield (cid, bytes);
+            // at this point, the leaf node may have max_degrees number of
+            // links, but no other stem node will
+        }
+
+        // our stream had 1 chunk that we have already yielded
+        if tree.len() == 1 && tree[0].len() == 1 {
+            return
+        }
+
+        // clean up, aka yield the rest of the stem nodes
+        // since all the stem nodes are able to recieve links
+        // we don't have to worry about "overflow"
+        while let Some(links) = tree.pop_front() {
+            // root node
+            // if tree.len() == 0 {
+            //     let (cid, bytes) = TreeNode::Root(links).encode()?;
+            //     yield (cid, bytes);
+            //     return
+            // }
+
+            let (cid, bytes) = TreeNode::Stem(links).encode()?;
+            let len = bytes.len();
+            yield (cid, bytes);
+
+            if let Some(front) = tree.front_mut() {
+                front.push((cid, len));
+            } else {
+                // final root, nothing to do
+            }
+        }
+    }
+}
+
+fn create_unixfs_node_from_links(links: Vec<(Cid, usize)>, is_root: bool) -> Result<UnixfsNode> {
+    let links = links
+        .into_iter()
+        .map(|(cid, len)| dag_pb::PbLink {
+            hash: Some(cid.to_bytes()),
+            // TODO: notes say that leaf chunks have "", does that mean that
+            // stem nodes should have None?
+            name: Some("".into()),
```

Comment:
the `unixfs` notes in question: https://www.notion.so/number-zero/UnixFs-742339892d9c47d5b79f4f942e661bbf#f74931c4d8b046c3807377cbe2b443f1

should I have `name` as `Some("".into())`, or `None`, or something else?

Reply:
yes, intermediary nodes should have no name, aka `None`.